File: frontend/src/api/apiKeys.ts
import { apiClient } from "@/api/client";

export type ApiKeyRecord = {
  id: string;
  name: string;
  key_prefix: string;
  is_active: boolean;
  last_used_at: string | null;
  created_at: string;
};

export type ApiKeyListResponse = {
  items: ApiKeyRecord[];
  total: number;
};

export type ApiKeyCreatedResponse = {
  id: string;
  name: string;
  key_prefix: string;
  api_key: string;
  created_at: string;
};

export async function listApiKeys(): Promise<ApiKeyListResponse> {
  const { data } = await apiClient.get<ApiKeyListResponse>("/api-keys");
  return data;
}

export async function createApiKey(name: string): Promise<ApiKeyCreatedResponse> {
  const { data } = await apiClient.post<ApiKeyCreatedResponse>("/api-keys", { name });
  return data;
}

export async function revokeApiKey(keyId: string): Promise<void> {
  await apiClient.delete(`/api-keys/${keyId}`);
}

File: backend/api/api_keys.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from backend.api.deps import get_current_user
from backend.database.session import get_db
from backend.models.user import User
from backend.schemas.api_key import ApiKeyCreateRequest, ApiKeyCreatedResponse, ApiKeyListResponse
from backend.services import api_key_service

router = APIRouter(prefix="/api-keys", tags=["api-keys"])


@router.get("", response_model=ApiKeyListResponse)
def list_keys(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    return api_key_service.list_api_keys(db, user)


@router.post("", response_model=ApiKeyCreatedResponse, status_code=status.HTTP_201_CREATED)
def create_key(
    body: ApiKeyCreateRequest,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return api_key_service.create_api_key(db, user, body.name)


@router.delete("/{key_id}", status_code=status.HTTP_204_NO_CONTENT)
def revoke_key(
    key_id: UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    if not api_key_service.revoke_api_key(db, user, key_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="API key not found")
